framework/src/meshgenerators/XYZDelaunayGenerator.C

```diff
@@ -0,0 +1,405 @@
+    return seed;
+  }
+};
+
+} // namespace std
+
+InputParameters
+XYZDelaunayGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+
+  MooseEnum algorithm("BINARY EXHAUSTIVE", "BINARY");
+
+  params.addRequiredParam<MeshGeneratorName>(
+      "boundary", "The input MeshGenerator defining the output outer boundary");
+
+  params.addParam<SubdomainName>("output_subdomain_name",
+                                 "Subdomain name to set on new triangles.");
+
+  params.addParam<BoundaryName>(
+      "output_boundary",
+      "Boundary name to set on new outer boundary.  Default ID: 0 if no hole meshes are stitched; "
+      "or maximum boundary ID of all the stitched hole meshes + 1.");
+  params.addParam<std::vector<BoundaryName>>(
+      "hole_boundaries",
+      "Boundary names to set on holes.  Default IDs are numbered up from 1 if no hole meshes are "
+      "stitched; or from maximum boundary ID of all the stitched hole meshes + 2.");
+
+  params.addParam<bool>("smooth_triangulation",
+                        false,
+                        "Whether to do Laplacian mesh smoothing on the generated triangles.");
+  params.addParam<std::vector<MeshGeneratorName>>(
+      "holes", std::vector<MeshGeneratorName>(), "The MeshGenerators that define mesh holes.");
+  params.addParam<std::vector<bool>>(
+      "stitch_holes", std::vector<bool>(), "Whether to stitch to the mesh defining each hole.");
+
+  params.addRangeCheckedParam<Real>(
+      "desired_volume",
+      0,
+      "desired_volume>=0",
+      "Desired (maximum) tetrahedral volume, or 0 to skip uniform refinement");
+
+  params.addParam<MooseEnum>(
+      "algorithm",
+      algorithm,
+      "Control the use of binary search for the nodes of the stitched surfaces.");
+  params.addParam<bool>(
+      "verbose_stitching", false, "Whether mesh stitching should have verbose output.");
+
+  params.addClassDescription(
+      "Creates tetrahedral 3D meshes within boundaries defined by input meshes.");
+
+  return params;
+}
+
+XYZDelaunayGenerator::XYZDelaunayGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _bdy_ptr(getMesh("boundary")),
+    _desired_volume(getParam<Real>("desired_volume")),
+    _output_subdomain_id(0),
+    _smooth_tri(getParam<bool>("smooth_triangulation")),
+    _hole_ptrs(getMeshes("holes")),
+    _stitch_holes(getParam<std::vector<bool>>("stitch_holes")),
+    _algorithm(parameters.get<MooseEnum>("algorithm")),
+    _verbose_stitching(parameters.get<bool>("verbose_stitching"))
+{
+  if (!_stitch_holes.empty() && _stitch_holes.size() != _hole_ptrs.size())
+    paramError("stitch_holes", "Need one stitch_holes entry per hole, if specified.");
+}
+
+std::unique_ptr<MeshBase>
+XYZDelaunayGenerator::generate()
+{
+#ifdef LIBMESH_HAVE_NETGEN
+  // Put the boundary mesh in a local pointer
+  std::unique_ptr<UnstructuredMesh> mesh =
+      dynamic_pointer_cast<UnstructuredMesh>(std::move(_bdy_ptr));
```

Comment:
I did some simples tests. The 3D meshes generated look great.
The mesh needs to be "prepared" here, otherwise the surface cannot be identified correctly.

Does this MG support 2D surface mesh (TRI3 elems) as input. I tried it and got an error.

Will do more tests and a more systematic review later this week.  Thanks.

Reply:
Thanks!

Yes, it needs neighbor pointers to be properly assigned ... I'll add a check for `!is_prepared()` and fix that up (probably at the libMesh level) when we find it's necessary.

What error did you get?  I've got a couple libMesh unit tests with Tri3 meshes as input.  They do also need neighbor pointers to have been assigned, *and* they need to be a closed manifold (every Tri3 needs to have neighbors on all sides).

Reply:
I double checked my 2D input mesh and found a bug. I fixed it and now it works perfectly. Thanks.

Reply:
I'd still like to know what error message you got.  If it pointed to the underlying bug then we're in great shape, if it pointed to some consequence of the underlying bug then we might be able to improve that, and if it pointed to something completely unrelated then we ought to fix that.

Reply:
I got a seg fault because the input mesh has extra TRI3 elements that do not belong to the surface.

Reply:
I get that you're the sort of user who can cope with my longstanding (bad) design philosophy of "As long as there's an assertion failure reporting a source file and line number then that should be all anyone needs to fix things, if not then that's a user Skill Issue" ... but MOOSE tries to be as friendly as possible to user errors even for less advanced users, and I'm trying to do better too.  :sweat_smile:

Reply:
A segfault is definitely unacceptable.  I'll see if I can replicate the problem.

Reply:
This should be resolved by https://github.com/libMesh/libmesh/pull/3919

Reply:
No, wait, the broken-boundary segfault was fixed, but the failure to handle unprepared meshes was not.  I'll get that now.

Reply:
Got that in libMesh/libmesh#3930